SimTransport/PPSProtonTransport/src/TotemTransport.cc

```diff
@@ -1,60 +1,62 @@
   m_sig_E = iConfig.getParameter<double>("BeamEnergyDispersion");
 
-  if (fPPSRegionStart_56 > 0)
-    fPPSRegionStart_56 *= -1;  // make sure sector 56 has negative position, as TOTEM convention
+  if (fPPSRegionStart_56_ > 0)
+    fPPSRegionStart_56_ *= -1;  // make sure sector 56 has negative position, as TOTEM convention
 
-  edm::LogInfo("TotemTransport") << "=============================================================================\n"
-                                 << "             Bulding LHC Proton transporter based on TOTEM model\n"
-                                 << "=============================================================================\n";
+  edm::LogVerbatim("TotemTransport")
+      << "=============================================================================\n"
+      << "             Bulding LHC Proton transporter based on TOTEM model\n"
+      << "=============================================================================\n";
 
-  m_aprox_ip_150_r = ReadParameterization(m_model_ip_150_r_name, beam1Filename_);
-  m_aprox_ip_150_l = ReadParameterization(m_model_ip_150_l_name, beam2Filename_);
+  m_aprox_ip_150_r_ = ReadParameterization(m_model_ip_150_r_name_, beam1Filename_);
+  m_aprox_ip_150_l_ = ReadParameterization(m_model_ip_150_l_name_, beam2Filename_);
 
-  if (m_aprox_ip_150_r == nullptr || m_aprox_ip_150_l == nullptr) {
-    edm::LogError("TotemTransport") << "Parameterisation " << m_model_ip_150_r_name << " or " << m_model_ip_150_l_name
+  if (m_aprox_ip_150_r_ == nullptr || m_aprox_ip_150_l_ == nullptr) {
+    edm::LogError("TotemTransport") << "Parameterisation " << m_model_ip_150_r_name_ << " or " << m_model_ip_150_l_name_
                                     << " missing in file. Cannot proceed. ";
-    exit(1);
+    throw edm::Exception(edm::errors::Configuration) << "TotemTransport is not properly initialized";
   }
-  edm::LogInfo("TotemTransport") << "Parameterizations read from file, pointers:" << m_aprox_ip_150_r << " "
-                                 << m_aprox_ip_150_l << " ";
+  edm::LogVerbatim("TotemTransport") << "Parameterizations read from file, pointers:" << m_aprox_ip_150_r_ << " "
+                                     << m_aprox_ip_150_l_ << " ";
 }
+TotemTransport::~TotemTransport() {}
 //
 // this method is the same for all propagator, but since transportProton is different for each derived class
 // it needes to be overriden
 //
-void TotemTransport::process(const HepMC::GenEvent* evt,
+void TotemTransport::process(const HepMC::GenEvent* ievt,
                              const edm::EventSetup& iSetup,
-                             CLHEP::HepRandomEngine* _engine) {
-  this->clear();
+                             CLHEP::HepRandomEngine* engine) {
+  clear();
+  engine_ = engine;  // the engine needs to be updated for each event
 
-  engine_ = _engine;  // the engine needs to be updated for each event
+  HepMC::GenEvent evt(*ievt);
```

Comment:
Just curious, IIUC the `transportProton()` modifies objects pointed to (and owned) by `evt`. How are the results returned (or communicated out) from this function?

Reply:
@makortel , we do the same in interface between generators and Geant4. Some parameters of HepMC data needs to be changed temporary in order to prepare a transport of scattered proton. This method allows to do that. After protons transport by Hector or Totem transports final proton at a distance of ~200 m from interaction point is added to Geant4 transport and provides hits in PPS/TOTEM detectors.

Reply:
Ah, I see, the results are communicated out via these member variables
https://github.com/cms-sw/cmssw/blob/43afc3b480eee8b8a865058b4c422b26ceaf86e3/SimTransport/PPSProtonTransport/src/TotemTransport.cc#L172-L174